SSLstrip - Adding the ability to choose how domains are spoofed

Petitoto opened the pull request:
This pull request is related to #723. I added the ability to choose how domains are spoofed.
Thus, I removed the `useIDN` option which was used to choose between two TLD modifications.

**How to choose how domains are spoofed?**
The new option is `http.proxy.sslstrip.replacements`.
It requires a space-separated list of `<original_domain>:<stripped_domain>`, ordered by priority. Internationalized Domain Names are supported. If the target domain isn't found in the list, sslstrip will double the last character of the domain name. Use `*` for any existing characters.<br />
For example :
`*.google.com:*.gooogle.com *.com:*.corn`
will replace *.google.com with *.gooogle.com (`www.google.com` -> `www.gooogle.com`), *.com with *.corn (`facebook.com` -> `facebook.corn`) but not `*.google.com` with `*.google.corn` (first rule takes precedence), and will double last char of the domain name for other websites (`wikipedia.org` -> `wikipedia.orgg`).<br />
Default values are : `*.com:*.corn *.org:*.orq *.net:*.nel`<br />
Note that if you want to erase the default behavior doubling the last character of the domain name, you can just add `*:*` at the end of the replacements.
For example :
`*.google.com:*.gooogle.com *:*`
will replace *.google.com with *.gooogle.com, but won't spoof anything else.
<br />
You can use multiple wildcards for more specific pattern matching. For example, `*.google.*:*.gooogle.*` is a valid parameter.

<br />

**Three additional fixes:**
- sslstrip couldn't handle 2 redirections (http://example.com -> https://example.com -> https://www.example.com) because it checked for a redirection from the HTTP scheme
- sslstrip used a long time-to-live (1024 seconds), which caused stripped hosts to be cached by target browsers even after the attack ended. Now it's reduced to 5 seconds (note that some browsers might ignore it)
- sslstrip couldn't edit subdomains to fix cookies

<br />
(Edited for the new commit)

buffermet commented:
@Petito how would we specify a subdomain from a TLD?

For example, how could we change all instances of `*.google.com` to `*.google.corn` without also spoofing `google.com.someotherhost.com`?

buffermet commented:
> If the target domain isn't found in the list, sslstrip will double the last character of the domain name.

I don't think this is a good way to default the routing. There are many cases where you don't want to affect the TLD of every domain.

Petitoto commented:
@buffermet 
>  how would we specify a subdomain from a TLD?
For example, how could we change all instances of *.google.com to *.google.corn without also spoofing *.google.com.someotherhost.com?

You can use the following settings: `google.com.:* google.com:google.corn`<br />
I understand that it's not an easy way to do this. However, it's sometimes interesting to spoof subdomains (when HSTS isn't loaded with `includeSubDomains` flag, when you just want to edit a specific pattern in any domain (e.g. google -> gooogle)...). Maybe another option specifying that sslstrip should spoof only hostnames / TLD could solve the issue. I'm not sure if that would be a good idea as adding more and more options will make using Bettercap more difficult.
<br />
> I don't think this is a good way to default the routing. There are many cases where you don't want to spoof the TLD of every domain.

The following settings will just spoof `com` to `corn` and will erase the default behavior: `com:corn *:*`<br />
I think that doubling the last character is interesting for a basic proof of the HSTS bypass.

buffermet commented:
@Petitoto 

> You can use the following settings: `google.com.:* google.com:google.corn`

Is this correct? Is there supposed to be a `com.`?

> I understand that it's not an easy way to do this. However, it's sometimes interesting to spoof subdomains (when HSTS isn't loaded with `includeSubDomains` flag, when you just want to edit a specific pattern in any domain (e.g. google -> gooogle)...). Maybe another option specifying that sslstrip should spoof only hostnames / TLD could solve the issue. I'm not sure if that would be a good idea as adding more and more options will make using Bettercap more difficult.

Using this configuration, I still don't see how the module would be able to tell a subdomain from a TLD. Look at the way I've implemented HSTSHIJACK. There is a format to which the module adheres. If you want to target all subdomains belonging to a specific domain, you use a prepended wildcard (i.e. `*.google.com`->`*.google.corn`)

Right now it seems you are splitting domains at every `.`.

> The following settings will just spoof `com` to `corn` and will erase the default behavior: `com:corn *:*`<br />
I think that doubling the last character is interesting for a basic proof of the HSTS bypass.

Again, there are many cases in which the attacker would want to target a specific domain, rather than all domains of a certain TLD.

Petitoto commented:
@buffermet 
It's correct, because domains to spoof are ordered by priority.
For example, using this settings : `google.com.:* google.com:gooogle.com *:*`
- sslstrip will first check for `google.com.` in target domains (i.e. google.com in a subdomain because of the last dot). If sslstrip find it, it will apply the rule which is to do nothing, otherwise it will try the next rule
- the next rule will check for `google.com` and replace it with `gooogle.com`
- finally, if `google.com` wasn't found, it will apply the last rule which is to do nothing for all others domains.

So sslstrip will strip `*google.com` to `*.gooogle.com`, won't spoof subdomains containing `google.com`, and won't spoof any other domain.
<br />
To make the use of sslstrip easier, I'll edit this pull request soon to be able to better differentiate TLD / subdomains using wildcards (something like in HSTSHIJACK).

buffermet commented:
@Petitoto looking at https://github.com/bettercap/bettercap/blob/a8f10b6333c81634ac557f44c939fa2d8f4251ec/modules/http_proxy/http_proxy_base_sslstriper.go#L217, this is not going to work.

You are doing a string replacement whereas spoofing hostnames within a header value is more complicated than when spoofing a hostname of an already parsed packet.

Take a look at https://github.com/bettercap/caplets/blob/master/hstshijack/hstshijack.js#L232 and https://github.com/bettercap/caplets/blob/master/hstshijack/hstshijack.js#L240

buffermet commented:
@Petitoto could I ask what's wrong with the hstshijack caplet?

I have never had any issues downgrading HSTS using that module. I understand that you were having problems with it, so it would be nice to get some feedback :)

Petitoto commented:
@buffermet 
I committed changes for a better use of wildcards (I updated my first message for new changes).

> ...that you are using LastIndex, so replacements like www.*:wvvw.* will behave differently. (i.e. www.*->wvvw.* would not work on domains like www.something.www.google.com)

You're right. It was to prioritize the occurence in the "highest level" of the domain in case of duplicates. Now it's using regex, so sslstrip will check exactly for `www.*` (starting with `www.` and ending with any characters), so it should be fixed.

> As far as I can follow the code, you are replacing in your replacement string every instance of * with the target string? Forgive me if I am missing things, but I do not see how in this case * could account for any and all subdomains, since you are splitting at every ..

It is no longer in the new commit. It was to replace `*` with any domain name, as wildcard had to be used alone.
However, I'm not splitting at every `.` for domain stripping.

> I would also suggest splitting fixCookies into 2 separate functions: spoofCookies and restoreCookies, that way it's easy to keep track of.

I'm not sure what you are suggesting. What would you like to put in restoreCookies? Something that would restore cookies to the original domain name? It would be more difficult because I think browsers won't accept cookies from another domain. SSLstrip should therefore handle one last connection to the spoofed domain before exitting, redirecting to the original domain name (this won't work on HSTS websites), and finally restore cookies.

> Also I am not sure why here you are splitting the domain at every ., and then only seem to target the final domain + TLD:

When setting cookies, domain name can be specified without subdomains in order to apply them on the whole domain (e.g. : `Set-Cookie: [...] domain=.google.com;`). Thus, it's required to split at each dot to check for domain to replace.
However, you're absolutely right: if a website specifies cookies on a specific subdomain and if this subdomain has been spoofed by sslstrip, only the hostname + TLD will be checked, so cookies may not work. I will try to fix this, thanks!

> could I ask what's wrong with the hstshijack caplet?

Sometimes I get dns issues, but I haven't investigated further. As it is quite rare, it could be dns cache issues and not because of hstshijack itself. I developed the sslstrip module not because hstshijack wasn't working, but rather for some functionalities (like cookies support, no javascript injection...), and because the one's included in Bettercap did not work :)

buffermet commented:
> > could I ask what's wrong with the hstshijack caplet?
> 
> Sometimes I get dns issues, but I haven't investigated further. As it is quite rare, it could be dns cache issues and not because of hstshijack itself. I developed the sslstrip module not because hstshijack wasn't working, but rather for some functionalities (like cookies support, no javascript injection...), and because the one's included in Bettercap did not work :)

What changes have you made to solve such DNS problems?

Bettercap has multiple methods with which you can inject JavaScript. HSTSHIJACK also uses a specific environment variable `hstshijack.payloads` which is used to inject JavaScript into your specified domains (wildcard is supported).

Also, cookie spoofing is already supported in hstshijack, and it's not limited to just cookies. It will also spoof any occurrence of a targeted hostname in **every** header, not just the cookies. You can see this [here](https://github.com/bettercap/caplets/blob/dba2ff08f0aed56ee81284a982ac9a04dec1c7c7/hstshijack/hstshijack.js#L589).

> I think browsers won't accept cookies from another domain. SSLstrip should therefore handle one last connection to the spoofed domain before exitting, redirecting to the original domain name (this won't work on HSTS websites), and finally restore cookies.

Cookies are domain specific, just like HSTS. If you think the objective of these types of attacks is to be able to facilitate an error-free Gmail session. Forget it.

The whole point of these downgrade attacks is to terminate the browsing session, and have the victim log in again. What happens from there on, is up to the attacker. This is why hstshijack includes the option to initiate a whitelisting callback from the victim's machine, letting you choose which hostnames should no longer be spoofed for said victim (e.g. after a successful phishing attack).

Petitoto commented:
@buffermet, I fixed the issue related to cookies & subdomain. Thanks for the report!

I'm sorry if I said any mistakes about hstshijack, I don't know it very well.

> Cookies are domain specific, just like HSTS. 

That's why I don't understand what would you like to put in "restoreCookies". I thought it was something which would restore the original domain name in cookies at the end of the attack, but I think I misunderstood what you said.

SSLstrip includes an option terminating the browsing session (I didn't code it, it was already in the module before I fixed it)

Does the new use of wildcards seem better to you?

evilsocket commented:
guys can you tell me at this point if this PR is ready to be merged?

Petitoto commented:
@evilsocket In my opinion, it's ready to be merged 👌

evilsocket commented:
@buffermet ?

buffermet commented:
@evilsocket, @Petitoto 
I don't think we should merge this.

Having to specify which domains we do not want to spoof is counter-intuitive, and impractical, take the following scenario:

I want to spoof: 
- `x.y`->`s.y`
- `x.y.z`->`s.y.z`

I do not want to spoof: 
- `x.y*`

There seems no way to achieve this with the proposed PR, because as described by @Petitoto, we would need to specify a replacement such as `x.y*:x.y*` and this would cause the desired replacements to fail. (think of domains such as `.co` and `.com`)

If this were to get patched, I still don't think we should implement domain spoofing like this, because we should be able to target a domain such as `*.google.com` without having to additionally fine-tune our scope by targeting domains that should **not** be spoofed, as the regexp is too permissive.

buffermet commented:
@Petitoto I now see [here](https://github.com/bettercap/bettercap/blob/b24d783f0b67214f3b23b78d0f821303025149fd/modules/http_proxy/http_proxy_base_sslstriper.go#L148) that you have implemented a whole search, which would fix the issue I mentioned, however, why are you splitting the replacement string at every "*" [here](https://github.com/bettercap/bettercap/blob/b24d783f0b67214f3b23b78d0f821303025149fd/modules/http_proxy/http_proxy_base_sslstriper.go#L151)?

Can you please confirm whether we no longer have to specify the domains that we don't want to target (i.e. `not.my.target.`->`not.my.target.`)?

If this is no longer an issue, I'd say we can merge this.

Petitoto commented:
@buffermet @evilsocket 
The issue you mentioned was fixed by the second commit. So now, you don't need to specify the domains you don't want to target. If you add `*.google.com`, it won't spoof `*.google.anotherdomain`.

> why are you splitting the replacement string at every *

You can use multiple `*` (eg: `*.google.*:*.gooogle.*`). Thus, the replacement string has to be changed from `*.gooogle.*` to `${1}.gooogle.${2}`, but it seems that there is no way in golang to do it automatically.

buffermet commented:
@Petitoto thanks for clarifying. I think splitting at every `*` is unnecessary, and the code could be shrunk slightly, but the only real issue I had with this PR was that you had to specify targets that you do not want to spoof.

I have not tested this yet, but I quite like this method of targeting hostnames. It seems like a simpler solution than having to align the `hstshijack.targets` and `hstshijack.replacements` variable values.

@evilsocket LGTM 👍

buffermet commented:
One more thing @Petitoto, you'll find that you won't **_need_** to support targets like `*.x.*`, because you couldn't pinpoint what host to spoof (or not). But it's still nice to have this option. 👍

buffermet commented:
I still think that we should remove this though: https://github.com/bettercap/bettercap/blob/b24d783f0b67214f3b23b78d0f821303025149fd/modules/http_proxy/http_proxy_base_sslstriper.go#L166

Or, if we are going this route (auto spoofing TLDs), we should implement a big warning to notify about what will happen, because technically speaking, SSLStripping is not the same as hostname spoofing.

Petitoto commented:
@buffermet I removed the default behavior doubling the last TLD's character.

Thanks in advance for testing :D

buffermet commented:
@Petitoto I have tested your PR.

It looks like the hostname spoofing works as expected, but nearly 99% of all sites break.

This is because you are only attacking on the proxy side. I still think you're better off using hstshijack and let the main payload do the spoofing once the page has successfully assembled. That is the best way to hijack the victim's navigation.

Petitoto commented:
Thanks @buffermet for your report.

What do you mean by "sites break"? I haven't any problem while sslstripping on the wide majority of websites.
Of course, for a few websites relying on specific and complex external resources, hstshijack might be able to produce better results. However, if you get into trouble for almost every websites, the issue must be because of another reason.

This issue isn't linked to this PR, but if you could share some details about this it would be interesting 👌.
So @evilsocket I think the PR is ready to be merged 👍

buffermet commented:
@Petitoto 

> This issue isn't linked to this PR

Well technically it is. You are not spoofing any HTML nodes using the proxy, you are only SSLStripping them.

Let's say you have spoofed mail.google.com to mail.google.corn, yet the resources on the page at mail.google.corn are loaded from mail.google.com.

If mail.google.com responds with an Access-Control-Allow-Origin header, it can restrict access to resources that were requested by certain hosts.

Since mail.google.com is HSTS preloaded, the site will break. In other words, the site will not render properly, scripts will not load, images will not load, and the page will not function.

If you really want to reinvent the wheel and achieve what hstshijack achieves, you COULD make the proxy spoof hostnames in the HTTP body, but that will create such an overhead, making the proxy even slower combined with SSLStrip, which you could avoid entirely by simply injecting JS, as is done with hstshijack.

buffermet commented:
@Petitoto why not implement your format of targeting hostnames in hstshijack?

It would be a great addition to the http module.

buffermet commented:
@Petitoto I must also apologise for making [this comment](https://github.com/bettercap/bettercap/pull/723#issuecomment-654065404):

> I'd like to see some of the optionality of hstshijack in the builtin proxy, for instance the ability to choose how a TLD/domain is spoofed (e.g. .com -> .corn), currently HTTP proxy modules lack the ability to make asynchronous calls, as well as making additional HTTP requests.

What I really meant to say was that I would like to have the ability to make asynchronous calls in scripts for the `http.proxy` module. Unfortunately the otto package does allow for asynchronous , nor does it include a `setTimeout()` method.

Making the proxy module spoof/strip response bodies is only going to add overhead to the proxy. We're better off making the victim's browser do all that work by injecting JavaScript.

I did not mean to confuse you and send you on a mission, sorry!

Petitoto commented:
@buffermet @evilsocket 

I don't understand why you closed the PR. This is just an improvment for my previous one. If you think that sslstrip module shouldn't be included in Bettercap, you should just remove it or revert my previous PR. If something in this PR doesn't match what you expected, just tell it.
I think that sslstrip module and hstshijack aren't in conflict as they behave differently. Sometimes, you would prefer a more stable sslstripping with JS injection and you'll use hstshijack, sometimes you would have to strip only from the proxy side, and you'll use the Bettercap module. Otherwise why was there already an sslstrip module + the hstshijack caplet?

I think that this PR should be merged, as it improves the current sslstrip and is working. I'm fully open to discussion in order to modify my code, and that's what I did 4 times.
I have something working on my computer: if I added this PR, this is for the Bettercap community.
Please, reopen it.
<br />
> Well technically it is. You are not spoofing any HTML nodes using the proxy, you are only SSLStripping them.

I still think the issue isn't linked to this PR. You'll have the same problem without these commits. This PR only allows users to have more control on how domains are spoofed.

> Since mail.google.com is HSTS preloaded, there is no way for us to intercept those response headers.

The website was intercepted, so sslstrip has stripped the links on the webpages, and external resources can be intercepted even if they are HSTS-preloaded.
There is one case for which sslstrip may not work: if external resources doesn't match the pattern of links to be replaced (i.e. if they are specified with the HTTP scheme but the website hosting them is HSTS preloaded (+ Access-Control-Allow-Origin header), which is a quite strange situation but could happen). How do hstshijack behave in this situation?

As I, and testers for my previous PR, have no problem while sslstripping, I still think there is another cause if 99% of websites you tested break. I could investigate more if you share details.
<br />
> why not implement your format of targeting hostnames in hstshijack?

For now it would be nice if this format of targeting hostnames was included in the sslstrip module...

buffermet commented:
> The website was intercepted, so sslstrip has stripped the links on the webpages, and external resources can be intercepted even if they are HSTS-preloaded.

That's not true, and I already explained why.

> There is one case for which sslstrip may not work: if external resources doesn't match the pattern of links to be replaced (i.e. if they are specified with the HTTP scheme but the website hosting them is HSTS preloaded (+ Access-Control-Allow-Origin header), which is a quite strange situation but could happen). How do hstshijack behave in this situation?

So you understand the problem and reason why I closed this PR. The hstshijack caplet both sslstrips and spoofs hostnames of URLs found in  `href` and `src` attributes in HTML documents. This is the only way to avoid breaking websites that are HSTS preloaded.

> This is just an improvment for my previous one.

If you were to remove the part that appends the `ノ ` to TLDs, the aforementioned issues would no longer occur, and the `http.proxy.sslstrip` module would be in a better state than it was thanks to the other SSLStripping improvements that you made.

Petitoto commented:
@buffermet
> The hstshijack caplet both sslstrips and spoofs hostnames of URLs found in href and src attributes in HTML documents

The sslstrip module should act in the same way for any HTTPS link in the webpage. If you got trouble with this, I would be happy to investigate :D

> So you understand the problem and reason why I closed this PR. 

?? This PR isn't about on how sslstrip strip resources but just add the ability to choose how domains are stripped

> If you were to remove the part that appends the ノ to TLDs, the aforementioned issues would no longer occur (i.e. websites will no longer break), and the http.proxy.sslstrip module would be in a better state than it was thanks to the other SSLStripping improvements that you made.

This is removed by this PR !?

buffermet commented:
> The sslstrip module should act in the same way for any HTTPS link in the webpage.

That was already the case in the first implementation of the sslstrip module.

> > So you understand the problem and reason why I closed this PR. 

> ?? This PR isn't about on how sslstrip strip resources but just add the ability to choose how domains are stripped

That's exactly the problem, I'm not going to keep repeating myself. You can investigate the issue by reading my comments.

Petitoto commented:
Hello @buffermet,
I would like to talk more about the issue you mentioned.

First, I'm really suprised of your results (`nearly 99% of all sites I visited break`). So I conducted few more tests, and here are some of my results:

![ExampleFirefox](https://user-images.githubusercontent.com/27863028/97109078-8927e080-16d1-11eb-8891-6c6d94dce3d3.PNG)
Link is correctly stripped according to the settings (`http` + `.orq`)

![VulnwebChrome](https://user-images.githubusercontent.com/27863028/97108930-e2434480-16d0-11eb-8e3b-cf7123558516.PNG)
<img src="https://user-images.githubusercontent.com/27863028/97108942-e8d1bc00-16d0-11eb-995b-07c5c24de254.PNG" width="400" height="200"><img src="https://user-images.githubusercontent.com/27863028/97108934-e40d0800-16d0-11eb-81dc-f116c82a311d.PNG" width="400" height="200">
Link is correctly stripped according to the settings and website doesn't break (the first image is the stripped host, the second is the normal one).

<img src="https://user-images.githubusercontent.com/27863028/97109077-888f4a00-16d1-11eb-9fbe-70a433236f11.png" width="200" height="400"><img src="https://user-images.githubusercontent.com/27863028/97109079-89c07700-16d1-11eb-8caa-01cf3b31b8c2.png" width="200" height="400"><img src="https://user-images.githubusercontent.com/27863028/97108933-e3747180-16d0-11eb-9ca0-ec73cab4ee76.png" width="200" height="400"><img src="https://user-images.githubusercontent.com/27863028/97108931-e2dbdb00-16d0-11eb-8344-757c1f877229.png" width="200" height="400">
Sites are correctly stripped and don't break on Firefox for Android

Stripped / Not stripped websites on various browsers:
<img src="https://user-images.githubusercontent.com/27863028/97108949-ec654300-16d0-11eb-9ad1-1d898d29776b.PNG" width="400" height="200"><img src="https://user-images.githubusercontent.com/27863028/97108939-e66f6200-16d0-11eb-9b6a-d4895c5246f8.PNG" width="400" height="200">
<img src="https://user-images.githubusercontent.com/27863028/97108926-e0798100-16d0-11eb-98b4-b763c4c2d823.PNG" width="400" height="200"><img src="https://user-images.githubusercontent.com/27863028/97108941-e8392580-16d0-11eb-9dfb-fb384e5f753f.PNG" width="400" height="200">
<img src="https://user-images.githubusercontent.com/27863028/97108945-eb341600-16d0-11eb-8e17-08e8dfda3b23.PNG" width="400" height="200"><img src="https://user-images.githubusercontent.com/27863028/97108936-e5d6cb80-16d0-11eb-9f83-f494119a03fb.PNG" width="400" height="200">
<img src="https://user-images.githubusercontent.com/27863028/97108943-e96a5280-16d0-11eb-9862-ab0ff5e56bfd.PNG" width="400" height="200"><img src="https://user-images.githubusercontent.com/27863028/97108935-e4a59e80-16d0-11eb-86a5-31ece300e8be.PNG" width="400" height="200">

So I don't get 99% of websites which break (note that the language of the browsers is french, but of course it doesn't impact the results).
Can you get the same, at least for these websites?

Here is the trouble I got sometimes:
- resources aren't loaded because of DNS issues: seems to be related to DNS cache issues + Bettercap too slow to send a DNS answer (especially whith a low-specs attacker + a lot of targets)
- a CORS issue, different from what you mentioned, when resources are loaded with the credential flag set:
![TwitterCORSErrorChrome](https://user-images.githubusercontent.com/27863028/97108927-e1121780-16d0-11eb-80c9-098e40cd0808.PNG)
- "specific" errors, for example when the stripped hostname is inside the url and used by the website:
![GoogleErrorFirefox](https://user-images.githubusercontent.com/27863028/97109075-87f6b380-16d1-11eb-8134-d09079f9a8a0.PNG)
(here, the problem is `continue=http://mail.google.corn` which should be `continue=http://mail.google.com`)

<br />
I don't understand this:

> Let's say you have spoofed mail.google.com to mail.google.corn, yet the resources on the page at mail.google.corn are loaded from mail.google.com.

> there is no way for us to set those response headers

> you COULD make the proxy spoof hostnames in the HTTP body

The module do strips the hostname of any https link in the HTTP body. That's why resources should be stripped, so they can be intercepted. (By the way, that's why the CORS error I mentioned happens: resource is well stripped & intercepted, headers are stripped, but when credentials are specified we must not add `Access-Control-Allow-Origin: *`).
It should behave the same as in #781: `stripURL` and `processURL` do basically the same thing (which is stripping links). Thus, if links are stripped using #781, they should be stripped using this PR (unless there is an error in `processURL`, but it would be easy to fix).

Using your example:
- target loads the resources at mail.google.corn because links in the HTTP body are stripped
- Bettercap loads the resources at mail.google.com
- Bettercap removes CORS headers and transmit the response to the target
- target doesn't see any CORS header, so it does not refuse the response

What I mentioned later (`There is one case for which sslstrip may not work: if external resources doesn't match the pattern of links to be replaced`) is when the resource is specified with the http scheme:
- link won't be stripped (because it's not an https link)
- if the host is HSTS preloaded, the browser will directly reach the website using the https scheme
- Bettercap can't intercept the website
- if the website specify `Access-Control-Allow-Origin`, the target browser will ignore the response

However, I never observed this during my tests, because `http`**~~`S`~~**`://HstsPreloadedWebsite.com` is counter-intuitive for web developers (so I think this is not the problem you encountered).

<br>

> that will create such an overhead, making the proxy even slower combined with SSLStrip, which you could avoid entirely by simply injecting JS, as is done with hstshijack.

When targeting a lot of targets with a low-specs attacker, Bettercap might get overloaded. In this situation, I totally agree with you: using hstshijack to let the target browser strip its own webpages is a better solution. However, for more basic use, proxy side stripping is working and more "silent".

<br>

> If you were to remove the part that appends the ノ to TLDs from your previous PR, the aforementioned issues would no longer occur

Here, I don't understand:
- why could internationalized domain names produce the issue you mentioned?
- removing `useIDN` is exactly what this PR do :sweat_smile:

If you still get broken websites, could you specify which ones & share the errors logs of your target browser?

buffermet commented:
> removing useIDN is exactly what this PR do 😅

This is not the only thing your PR does, I created a PR and merged it which **specifically** removes the useIDN thing.

You also keep saying that there is "one" scenario where this method won't work, and then continue to list several...

It seems you are well aware of why this is a bad idea, and now I'll explain it to you for a third time:

Spoofing and stripping domains on the proxy side is a bad idea because of the overhead it creates. In other words, this is only gonna slow bettercap down without giving us any additional advantage, while you can simply spoof headers and inject payloads for the victim's browser to execute (in order to spoof the page). But since your PR does not do this (stripping AND spoofing on the proxy site), a lot of scripts and other resources will be blocked due to invalid/missing Access-Control-Allow-Origin headers (which are controlled by the responding servers).

Petitoto commented:
@buffermet 
What I explained is that you should not get any Access-Control-Allow-Origin issue for the wide majority of websites.

Stripping the http body from the proxy side doesn't slow down bettercap for basic use.
Currently, with your PR, you still strip the body from the proxy side in order to replace https links to http ones (moreover, you haven't removed the host tracker & the cookie restore, so there is almost the same overhead).

I think we should keep a proxy-side stripping in Bettercap, because:
- hstshijack already exists, so I don't think implementing something similar in Bettercap would be useful
- proxy-side stripping can sometimes be an advantage (more discreet)
- proxy modules are more meant to be proxy-side

What do you think if we disable domain spoofing by default, with a better use of the host tracker so that we don't slow down Bettercap, but keep the ability to spoof hostnames from the proxy side if the user wants to?

buffermet commented:
> Stripping the http body from the proxy side doesn't slow down bettercap for basic use.

Maybe if you're testing locally, but not when you're MITMing a network and get multiple HTTP requests per second...

> Currently, with your PR, you still strip the body from the proxy side in order to replace https links to http ones (moreover, you haven't removed the host tracker & the cookie restore, so there is almost the same overhead).

You are mixing 2 PRs now which is confusing. The way SSLStrip works now is how it is supposed to. It strips the `s` from `https`.

> I think we should keep a proxy-side stripping in Bettercap

It already exists. I am telling you that searching every HTTP body for every targeted hostname to spoof is a bad idea.

> hstshijack already exists, so I don't think implementing something similar in Bettercap would be useful

True, especially for the technical reasons I have mention multiple times here.

> proxy modules are more meant to be proxy-side

I don't know what you're saying here, because hstshijack is a HTTP proxy script.

<hr>

This is the last time that I am explaining to you why this is a bad idea.

buffermet commented:
Lastly:

> What I explained is that you should not get any Access-Control-Allow-Origin issue for the wide majority of websites.

You will get this error for most (if not all) websites that try to serve resources requested via HTTPS from a spoofed document.
